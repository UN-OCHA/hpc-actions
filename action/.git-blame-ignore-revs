021360ee3e3112996aa2b003577694984ede31fa
58d30143a6eec524084f39a1c6233acbed4fe0f2
c3611d633129d194471d09bfdcec1d6fdb042828
d8aacaa7ae644cbaf1130bee3286a2883a9c8427
