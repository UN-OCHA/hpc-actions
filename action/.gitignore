/node_modules
/lib
/.coverage
/test/data
/coverage.lcov
/.env