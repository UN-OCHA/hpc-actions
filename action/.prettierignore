/node_modules
/lib
/test/data
