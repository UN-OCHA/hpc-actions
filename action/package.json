{
  "name": "hpc-actions",
  "scripts": {
    "build": "tsc && ncc build lib/src",
    "test": "jest --projects test/jest.config.js --runInBand --verbose"
  },
  "engines": {
    "node": ">=16.15.0",
    "npm": ">=8.10.0"
  },
  "devDependencies": {
    "@octokit/webhooks-types": "^6.3.5",
    "@types/jest": "^28.1.7",
    "@types/node": "^18.7.6",
    "@vercel/ncc": "^0.34.0",
    "jest": "^28.1.3",
    "nyc": "^15.1.0",
    "ts-jest": "^28.0.8",
    "typescript": "^4.7.4"
  },
  "dependencies": {
    "@octokit/rest": "^18.0.0",
    "fp-ts": "^2.6.5",
    "io-ts": "^2.2.5",
    "isomorphic-git": "^1.8.2"
  }
}
