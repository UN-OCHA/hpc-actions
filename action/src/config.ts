/**
 * Configuration for this action.
 * 
 * Rather than provide all configuration options via environment variables,
 * we've opted to use json configuration files instead to allow for more
 * extensibility. This configuration file will be used for all options,
 * except for secrets, and the config file pathname.
 */
import * as t from 'io-ts';
import { PathReporter } from 'io-ts/lib/PathReporter';
import { either } from 'fp-ts';
import { promises as fs } from 'fs';

/**
 * The environment variables that are expected
 */

export interface Env {
  /**
   * The path to the configuration json file,
   * that adheres to the specification set out by CONFIG above.
   */
  CONFIG_FILE?: string;
  /**
   * The username to login to the docker registry with
   */
  DOCKER_USERNAME?: string;
  /**
   * The password to login to the docker registry with
   */
  DOCKER_PASSWORD?: string;
  /**
   * Token to use when interacting with the GitHub API
   */
  GITHUB_TOKEN?: string;

  // Implicit environment variables passed by GitHub
  GITHUB_REPOSITORY?: string;
  GITHUB_EVENT_NAME?: string;
  GITHUB_EVENT_PATH?: string;
  GITHUB_ACTOR?: string;
}

// Specify the configuration options using io-ts,
// which provides both type-definitions for the configuration,
// and validation that matches these definitions.

const DOCKER_CONFIG = t.intersection([
  // Required Config
  t.type({
    /**
     * Where in the repository should the build be run from
     */
    path: t.string,
    /**
     * What are the names of the build arguments that the docker image
     * expects to be supplied
     */
    args: t.type({
      /**
       * What is the name of the build argument that expects the commit sha
       */
      commitSha: t.string,
      /**
       * What is the name of the build argument that expects the tree sha
       */
      treeSha: t.string,
    }),
    /**
     * What are the names of the environment variables where important bits of
     * information are stored
     */
    environmentVariables: t.type({
      /**
       * What environment variable is used to store the commit sha
       */
      commitSha: t.string,
      /**
       * What environment variable is used to store the tree sha
       */
      treeSha: t.string,
    }),
    /**
     * What's the name of the repository that we'll be tagging
     */
    repository: t.string,
  }),
  // Optional config
  t.partial({
    /**
     * If provided, use the given registry instead of Docker Hub.
     *
     * When this is set, the docker repository must start with this string
     * followed by a slash.
     */
    registry: t.string,
  })
]);

const CONFIG = t.intersection([
  // Required Config
  t.type({
    /**
     * What is the branch for the staging environment.
     */
    stagingEnvironmentBranch: t.keyof({
      'env/stage': null,
      'env/staging': null,
    }),
    /**
     * What are the branches used to track development environments.
     */
    developmentEnvironmentBranches: t.array(t.string),
    /**
     * What type of repository is this?
     * 
     * This is used to work out how to calculate the current version
     */
    repoType: t.keyof({
      /**
       * Calculate the current version by reading the version from package.json
       */
      'node': null,
    }),
    /**
     * Configuration for the docker image build and publication
     */
    docker: DOCKER_CONFIG,
    /**
     * List of shell commands to run as part of CI
     */
    ci: t.array(t.string),
  }),
  // Optional config
  t.partial({
    /**
     * If provided, add these labels to mergeback pull requests
     */
    mergebackLabels: t.array(t.string),
  })
]);

/**
 * The type of a valid configuration file
 */
export type Config = t.TypeOf<typeof CONFIG>;

/**
 * The type for valid docker configuration
 */
export type DockerConfig = t.TypeOf<typeof DOCKER_CONFIG>;

export const getConfig = async (env: Env): Promise<Config> => {
  if (!env.CONFIG_FILE) {
    throw new Error('Environment Variable CONFIG_FILE is required');
  }
  const data = await fs.readFile(env.CONFIG_FILE)
    .catch(err => {
      if (err.message?.startsWith('ENOENT: no such file or directory')) {
        throw new Error(`Could not find configuration file "${env.CONFIG_FILE}" specified in CONFIG_FILE`);
      } else {
        throw err;
      }
    });
  let json: string;
  try {
    json = JSON.parse(data.toString())
  } catch (err) {
    throw new Error(`The configuration file at "${env.CONFIG_FILE}" is not valid JSON: ${err.message}`);
  }
  const config = CONFIG.decode(json);
  if (either.isLeft(config)) {
    throw new Error('Invalid Configuration: \n* ' + PathReporter.report(config).join('\n* '));
  }
  // Check that development environments all start with `env/`
  for (const env of config.right.developmentEnvironmentBranches) {
    if (!env.startsWith('env/')) {
      throw new Error('Invalid Configuration: All development environment branches must start with env/');
    }
  }
  if (
    config.right.docker.registry &&
    !config.right.docker.repository.startsWith(`${config.right.docker.registry}/`)
  ) {
    throw new Error(
      'Invalid Configuration: Docker repository must start with: ' +
      `${config.right.docker.registry}/`
    );
  }
  return config.right;
};
