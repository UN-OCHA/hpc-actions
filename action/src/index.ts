import { runAction } from './action';
import type { Env } from './config';

runAction({
  env: process.env as unknown as Env,
}).catch(async (err) => {
  console.log(`##[error] ${err.message}`);
  setTimeout(() => {
    console.error(err);
    process.exit(1);
  }, 100);
});
