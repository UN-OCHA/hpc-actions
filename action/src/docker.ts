import * as t from 'io-ts';
import { PathReporter } from 'io-ts/lib/PathReporter';
import { isLeft } from 'fp-ts/lib/Either';

import { exec, execAndPipeOutput } from './util/child_process';
import { Logger } from './util/interfaces';

import { DockerConfig } from './config';

export type DockerInit = (config: DockerConfig) => DockerController;

export interface DockerImageMetadata {
  commitSha: string;
  treeSha: string;
}

export interface DockerController {
  login: (opts: {
    user: string;
    pass: string;
  }) => Promise<void>;
  /**
   * Get the metadata for a docker image that is tagged locally
   */
  getMetadata: (tag: string) => Promise<DockerImageMetadata>;
  /**
   * Try and pull a docker image with the given tag,
   * return true if successful and false if not
   */
  pullImage: (tag: string, logger: Logger) => Promise<boolean>;
  /**
   * Run the docker build, and tag the image with the given tag
   */
  runBuild: (opts: {
    cwd: string;
    tag: string;
    meta: DockerImageMetadata;
    logger: Logger;
  }) => Promise<void>;
  /**
   * Push the image with the given tag to the configured destination
   */
  pushImage: (tag: string) => Promise<void>;
}

/**
 * Codec to consume environment variables from a docker image
 */
const IMAGE_DETAILS = t.array(t.type({
  Config: t.type({
    Env: t.array(t.string)
  })
}));

export const REAL_DOCKER: DockerInit = config => ({

  login: async ({user, pass}) => {
    // Login to docker
    await execAndPipeOutput({
      command: `docker login ${config.registry || ''} -u ${user}  --password-stdin`,
      cwd: __dirname,
      // Drop all console output (it's mostly warning about storing credentials)
      logger: console,
      data: pass + '\n'
    });
  },

  pullImage: (tag, logger) =>
    execAndPipeOutput({
      command: `docker pull ${config.repository}:${tag}`,
      cwd: __dirname,
      logger
    }).then(() => true).catch(() => false),

  getMetadata: async tag => {
    const res = await exec(`docker inspect ${config.repository}:${tag}`);
    const data = JSON.parse(res.stdout);
    const check = IMAGE_DETAILS.decode(data);
    if (isLeft(check)) {
      throw new Error(
        'Unexpected output from docker inspect: \n* ' +
        PathReporter.report(check).join('\n* ')
      );
    }
    if (check.right.length !== 1) {
      throw new Error('Unexpected output from docker inspect: multiple objects');
    }
    const image = check.right[0];
    // Able to parse output
    let commitSha: string | null = null;
    let treeSha: string | null = null;
    const varConfig = config.environmentVariables;
    for (const envVar of image.Config.Env) {
      if (envVar.startsWith(`${varConfig.commitSha}=`)) {
        commitSha = envVar.substr(varConfig.commitSha.length + 1);
      }
      if (envVar.startsWith(`${varConfig.treeSha}=`)) {
        treeSha = envVar.substr(varConfig.treeSha.length + 1);
      }
    }
    if (!commitSha || !treeSha) {
      throw new Error(
        'Unable to extract treeSha and commitSha from docker image'
      );
    }
    return {
      commitSha,
      treeSha,
    };
  },

  runBuild: async ({ cwd, tag, meta, logger }) => {
    await execAndPipeOutput({
      command: (
        `docker build ${config.path} ` +
        `--build-arg ${config.args.commitSha}=${meta.commitSha} ` +
        `--build-arg ${config.args.treeSha}=${meta.treeSha} ` +
        `-t ${config.repository}:${tag}`
      ),
      logger,
      cwd
    })
  },

  pushImage: tag =>
    exec(`docker push ${config.repository}:${tag}`).then(() => {}),

});
