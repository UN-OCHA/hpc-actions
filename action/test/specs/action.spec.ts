import * as action from '../../src/action';

describe('action', () => {

  describe('runAction', () => {
    it('Hello World', () => {
      expect(action.runAction()).toEqual('hello world');
    });
  });

});
