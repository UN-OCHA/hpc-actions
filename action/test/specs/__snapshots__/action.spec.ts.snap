// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action runAction push to production or staging env/prod ci Check directory 1`] = `
Array [
  Array [
    "[stdout]",
    "##[info] Handling push to branch env/prod",
  ],
  Array [
    "[stdout]",
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Logging in to docker",
  ],
  Array [
    "[stdout]",
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Image with tag v1.2.0 does not yet exist, building image",
  ],
  Array [
    "[stdout]",
    "##[info] Image built, checking tag is unchanged",
  ],
  Array [
    "[stdout]",
    "##[info] Tag is unchanged, okay to continue",
  ],
  Array [
    "[stdout]",
    "##[info] Pushing image to docker repository",
  ],
  Array [
    "[stdout]",
    "##[info] Image Pushed",
  ],
  Array [
    "[stdout]",
    "##[info] Running CI Checks",
  ],
  Array [
    "[stdout]",
    "##[info] Running: cat package.json",
  ],
  Array [
    "[stdout]",
    "{\\"version\\":\\"1.2.0\\"}",
  ],
  Array [
    "[stdout]",
    "##[info] CI Checks Complete",
  ],
  Array [
    "[stdout]",
    "##[info] Creating and pushing mergeback Branch: mergeback/prod/1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Opening Mergeback Pull Request",
  ],
]
`;

exports[`action runAction push to production or staging env/prod ci Command with interleaved stderr and stdout 1`] = `
Array [
  Array [
    "[stdout]",
    "##[info] Handling push to branch env/prod",
  ],
  Array [
    "[stdout]",
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Logging in to docker",
  ],
  Array [
    "[stdout]",
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Image with tag v1.2.0 does not yet exist, building image",
  ],
  Array [
    "[stdout]",
    "##[info] Image built, checking tag is unchanged",
  ],
  Array [
    "[stdout]",
    "##[info] Tag is unchanged, okay to continue",
  ],
  Array [
    "[stdout]",
    "##[info] Pushing image to docker repository",
  ],
  Array [
    "[stdout]",
    "##[info] Image Pushed",
  ],
  Array [
    "[stdout]",
    "##[info] Running CI Checks",
  ],
  Array [
    "[stdout]",
    "##[info] Running: echo && echo foo && sleep 0.1s && echo bar 1>&2 && sleep 0.1s && echo baz",
  ],
  Array [
    "[stdout]",
    "",
  ],
  Array [
    "[stdout]",
    "foo",
  ],
  Array [
    "[stderr]",
    "bar",
  ],
  Array [
    "[stdout]",
    "baz",
  ],
  Array [
    "[stdout]",
    "##[info] CI Checks Complete",
  ],
  Array [
    "[stdout]",
    "##[info] Creating and pushing mergeback Branch: mergeback/prod/1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Opening Mergeback Pull Request",
  ],
]
`;

exports[`action runAction push to production or staging env/prod ci Command with nonzero exit code 1`] = `
Array [
  Array [
    "[stdout]",
    "##[info] Handling push to branch env/prod",
  ],
  Array [
    "[stdout]",
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Logging in to docker",
  ],
  Array [
    "[stdout]",
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Image with tag v1.2.0 does not yet exist, building image",
  ],
  Array [
    "[stdout]",
    "##[info] Image built, checking tag is unchanged",
  ],
  Array [
    "[stdout]",
    "##[info] Tag is unchanged, okay to continue",
  ],
  Array [
    "[stdout]",
    "##[info] Pushing image to docker repository",
  ],
  Array [
    "[stdout]",
    "##[info] Image Pushed",
  ],
  Array [
    "[stdout]",
    "##[info] Running CI Checks",
  ],
  Array [
    "[stdout]",
    "##[info] Running: exit 123",
  ],
]
`;

exports[`action runAction push to production or staging env/prod docker Existing Image (different sha) 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/prod",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
  Array [
    "##[info] Image already exists, checking it was built with same git tree",
  ],
]
`;

exports[`action runAction push to production or staging env/prod docker Existing Image (different sha) 2`] = `
Array [
  Array [
    "v1.2.0",
  ],
]
`;

exports[`action runAction push to production or staging env/prod docker Existing Image (different sha) 3`] = `
Array [
  Array [
    "v1.2.0",
  ],
]
`;

exports[`action runAction push to production or staging env/prod docker Existing Image (matching sha) 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/prod",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
]
`;

exports[`action runAction push to production or staging env/prod docker Existing Image (matching sha) 2`] = `Array []`;

exports[`action runAction push to production or staging env/prod docker Non-Existant Image (tag changed) 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/prod",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
  Array [
    "##[info] Image with tag v1.2.0 does not yet exist, building image",
  ],
  Array [
    "##[info] Image built, checking tag is unchanged",
  ],
]
`;

exports[`action runAction push to production or staging env/prod docker Non-Existant Image (tag changed) 2`] = `
Object {
  "getMetadata": Array [],
  "pullImage": Array [
    Array [
      "v1.2.0",
    ],
  ],
  "pushImage": Array [],
}
`;

exports[`action runAction push to production or staging env/prod docker Non-Existant Image 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/prod",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
  Array [
    "##[info] Image with tag v1.2.0 does not yet exist, building image",
  ],
  Array [
    "##[info] Image built, checking tag is unchanged",
  ],
  Array [
    "##[info] Tag is unchanged, okay to continue",
  ],
  Array [
    "##[info] Pushing image to docker repository",
  ],
  Array [
    "##[info] Image Pushed",
  ],
  Array [
    "##[info] Running CI Checks",
  ],
  Array [
    "##[info] CI Checks Complete",
  ],
  Array [
    "##[info] Creating and pushing mergeback Branch: mergeback/prod/1.2.0",
  ],
  Array [
    "##[info] Opening Mergeback Pull Request",
  ],
]
`;

exports[`action runAction push to production or staging env/prod docker Non-Existant Image 2`] = `
Object {
  "getMetadata": Array [],
  "pullImage": Array [
    Array [
      "v1.2.0",
    ],
  ],
  "pushImage": Array [
    Array [
      "v1.2.0",
    ],
  ],
}
`;

exports[`action runAction push to production or staging env/prod mergeback 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/prod",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
  Array [
    "##[info] Image with tag v1.2.0 does not yet exist, building image",
  ],
  Array [
    "##[info] Image built, checking tag is unchanged",
  ],
  Array [
    "##[info] Tag is unchanged, okay to continue",
  ],
  Array [
    "##[info] Pushing image to docker repository",
  ],
  Array [
    "##[info] Image Pushed",
  ],
  Array [
    "##[info] Running CI Checks",
  ],
  Array [
    "##[info] CI Checks Complete",
  ],
  Array [
    "##[info] Creating and pushing mergeback Branch: mergeback/prod/1.2.0",
  ],
  Array [
    "##[info] Opening Mergeback Pull Request",
  ],
  Array [
    "##[info] Pull Request Opened, workflow complete",
  ],
]
`;

exports[`action runAction push to production or staging env/prod mergeback 2`] = `
Array [
  Array [
    Object {
      "base": "env/staging",
      "head": "mergeback/prod/1.2.0",
      "labels": Array [],
      "title": "Update env/staging with changes from env/prod",
    },
  ],
]
`;

exports[`action runAction push to production or staging env/prod tagging Existing tag (changed tree) 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/prod",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] The tag v1.2.0 already exists, checking that tree hasn't changed",
  ],
]
`;

exports[`action runAction push to production or staging env/prod tagging Existing tag (current commit) 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/prod",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] The tag v1.2.0 already exists, checking that tree hasn't changed",
  ],
  Array [
    "##[info] The tag is for the current commit, okay to continue",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
]
`;

exports[`action runAction push to production or staging env/prod tagging Existing tag (different commit, matching tree) 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/prod",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] The tag v1.2.0 already exists, checking that tree hasn't changed",
  ],
  Array [
    "##[info] The current tree matches the existing tag, okay to continue",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
]
`;

exports[`action runAction push to production or staging env/prod tagging Non-Existant Tag 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/prod",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
]
`;

exports[`action runAction push to production or staging env/staging ci Check directory 1`] = `
Array [
  Array [
    "[stdout]",
    "##[info] Handling push to branch env/staging",
  ],
  Array [
    "[stdout]",
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Logging in to docker",
  ],
  Array [
    "[stdout]",
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Image with tag v1.2.0 does not yet exist, building image",
  ],
  Array [
    "[stdout]",
    "##[info] Image built, checking tag is unchanged",
  ],
  Array [
    "[stdout]",
    "##[info] Tag is unchanged, okay to continue",
  ],
  Array [
    "[stdout]",
    "##[info] Pushing image to docker repository",
  ],
  Array [
    "[stdout]",
    "##[info] Image Pushed",
  ],
  Array [
    "[stdout]",
    "##[info] Running CI Checks",
  ],
  Array [
    "[stdout]",
    "##[info] Running: cat package.json",
  ],
  Array [
    "[stdout]",
    "{\\"version\\":\\"1.2.0\\"}",
  ],
  Array [
    "[stdout]",
    "##[info] CI Checks Complete",
  ],
  Array [
    "[stdout]",
    "##[info] Creating and pushing mergeback Branch: mergeback/staging/1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Opening Mergeback Pull Request",
  ],
]
`;

exports[`action runAction push to production or staging env/staging ci Command with interleaved stderr and stdout 1`] = `
Array [
  Array [
    "[stdout]",
    "##[info] Handling push to branch env/staging",
  ],
  Array [
    "[stdout]",
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Logging in to docker",
  ],
  Array [
    "[stdout]",
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Image with tag v1.2.0 does not yet exist, building image",
  ],
  Array [
    "[stdout]",
    "##[info] Image built, checking tag is unchanged",
  ],
  Array [
    "[stdout]",
    "##[info] Tag is unchanged, okay to continue",
  ],
  Array [
    "[stdout]",
    "##[info] Pushing image to docker repository",
  ],
  Array [
    "[stdout]",
    "##[info] Image Pushed",
  ],
  Array [
    "[stdout]",
    "##[info] Running CI Checks",
  ],
  Array [
    "[stdout]",
    "##[info] Running: echo && echo foo && sleep 0.1s && echo bar 1>&2 && sleep 0.1s && echo baz",
  ],
  Array [
    "[stdout]",
    "",
  ],
  Array [
    "[stdout]",
    "foo",
  ],
  Array [
    "[stderr]",
    "bar",
  ],
  Array [
    "[stdout]",
    "baz",
  ],
  Array [
    "[stdout]",
    "##[info] CI Checks Complete",
  ],
  Array [
    "[stdout]",
    "##[info] Creating and pushing mergeback Branch: mergeback/staging/1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Opening Mergeback Pull Request",
  ],
]
`;

exports[`action runAction push to production or staging env/staging ci Command with nonzero exit code 1`] = `
Array [
  Array [
    "[stdout]",
    "##[info] Handling push to branch env/staging",
  ],
  Array [
    "[stdout]",
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Logging in to docker",
  ],
  Array [
    "[stdout]",
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
  Array [
    "[stdout]",
    "##[info] Image with tag v1.2.0 does not yet exist, building image",
  ],
  Array [
    "[stdout]",
    "##[info] Image built, checking tag is unchanged",
  ],
  Array [
    "[stdout]",
    "##[info] Tag is unchanged, okay to continue",
  ],
  Array [
    "[stdout]",
    "##[info] Pushing image to docker repository",
  ],
  Array [
    "[stdout]",
    "##[info] Image Pushed",
  ],
  Array [
    "[stdout]",
    "##[info] Running CI Checks",
  ],
  Array [
    "[stdout]",
    "##[info] Running: exit 123",
  ],
]
`;

exports[`action runAction push to production or staging env/staging docker Existing Image (different sha) 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/staging",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
  Array [
    "##[info] Image already exists, checking it was built with same git tree",
  ],
]
`;

exports[`action runAction push to production or staging env/staging docker Existing Image (different sha) 2`] = `
Array [
  Array [
    "v1.2.0",
  ],
]
`;

exports[`action runAction push to production or staging env/staging docker Existing Image (different sha) 3`] = `
Array [
  Array [
    "v1.2.0",
  ],
]
`;

exports[`action runAction push to production or staging env/staging docker Existing Image (matching sha) 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/staging",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
]
`;

exports[`action runAction push to production or staging env/staging docker Existing Image (matching sha) 2`] = `Array []`;

exports[`action runAction push to production or staging env/staging docker Non-Existant Image (tag changed) 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/staging",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
  Array [
    "##[info] Image with tag v1.2.0 does not yet exist, building image",
  ],
  Array [
    "##[info] Image built, checking tag is unchanged",
  ],
]
`;

exports[`action runAction push to production or staging env/staging docker Non-Existant Image (tag changed) 2`] = `
Object {
  "getMetadata": Array [],
  "pullImage": Array [
    Array [
      "v1.2.0",
    ],
  ],
  "pushImage": Array [],
}
`;

exports[`action runAction push to production or staging env/staging docker Non-Existant Image 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/staging",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
  Array [
    "##[info] Image with tag v1.2.0 does not yet exist, building image",
  ],
  Array [
    "##[info] Image built, checking tag is unchanged",
  ],
  Array [
    "##[info] Tag is unchanged, okay to continue",
  ],
  Array [
    "##[info] Pushing image to docker repository",
  ],
  Array [
    "##[info] Image Pushed",
  ],
  Array [
    "##[info] Running CI Checks",
  ],
  Array [
    "##[info] CI Checks Complete",
  ],
  Array [
    "##[info] Creating and pushing mergeback Branch: mergeback/staging/1.2.0",
  ],
  Array [
    "##[info] Opening Mergeback Pull Request",
  ],
]
`;

exports[`action runAction push to production or staging env/staging docker Non-Existant Image 2`] = `
Object {
  "getMetadata": Array [],
  "pullImage": Array [
    Array [
      "v1.2.0",
    ],
  ],
  "pushImage": Array [
    Array [
      "v1.2.0",
    ],
  ],
}
`;

exports[`action runAction push to production or staging env/staging mergeback 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/staging",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
  Array [
    "##[info] Image with tag v1.2.0 does not yet exist, building image",
  ],
  Array [
    "##[info] Image built, checking tag is unchanged",
  ],
  Array [
    "##[info] Tag is unchanged, okay to continue",
  ],
  Array [
    "##[info] Pushing image to docker repository",
  ],
  Array [
    "##[info] Image Pushed",
  ],
  Array [
    "##[info] Running CI Checks",
  ],
  Array [
    "##[info] CI Checks Complete",
  ],
  Array [
    "##[info] Creating and pushing mergeback Branch: mergeback/staging/1.2.0",
  ],
  Array [
    "##[info] Opening Mergeback Pull Request",
  ],
  Array [
    "##[info] Pull Request Opened, workflow complete",
  ],
]
`;

exports[`action runAction push to production or staging env/staging mergeback 2`] = `
Array [
  Array [
    Object {
      "base": "develop",
      "head": "mergeback/staging/1.2.0",
      "labels": Array [],
      "title": "Update develop with changes from env/staging",
    },
  ],
]
`;

exports[`action runAction push to production or staging env/staging tagging Existing tag (changed tree) 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/staging",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] The tag v1.2.0 already exists, checking that tree hasn't changed",
  ],
]
`;

exports[`action runAction push to production or staging env/staging tagging Existing tag (current commit) 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/staging",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] The tag v1.2.0 already exists, checking that tree hasn't changed",
  ],
  Array [
    "##[info] The tag is for the current commit, okay to continue",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
]
`;

exports[`action runAction push to production or staging env/staging tagging Existing tag (different commit, matching tree) 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/staging",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] The tag v1.2.0 already exists, checking that tree hasn't changed",
  ],
  Array [
    "##[info] The current tree matches the existing tag, okay to continue",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
]
`;

exports[`action runAction push to production or staging env/staging tagging Non-Existant Tag 1`] = `
Array [
  Array [
    "##[info] Handling push to branch env/staging",
  ],
  Array [
    "##[info] Checking if there is an existing tag for v1.2.0",
  ],
  Array [
    "##[info] Creating and pushing new tag v1.2.0",
  ],
  Array [
    "##[info] Logging in to docker",
  ],
  Array [
    "##[info] Checking for existing docker image with tag v1.2.0",
  ],
]
`;
